### PR TITLE
videodemuxer: added seek functionality

### PR DESCRIPTION
**Pull request by rrawther**
added seek functionality to video demuxer

### CODE REVIEW

File: utils/video_demuxer.h [hunk 1]
```diff
@@ -40,6 +40,84 @@ extern "C" {
  * \brief AMD The rocDecode video demuxer API.
  */
 
+/**
+ * @brief Enum for Seek mode
+ * 
+ */
+typedef enum Seekseek_mode_enum {
```
Comment (AryanSalmanpour) [inline utils/video_demuxer.h]: change `Seekseek_mode_enum` to `SeekModeEnum`

File: utils/video_demuxer.h [hunk 2]
```diff
@@ -40,6 +40,84 @@ extern "C" {
...
+    SEEK_MODE_EXACT_FRAME = 0,
+    SEEK_MODE_PREV_KEY_FRAME = 1,
+    SEEK_MODE_NUM,
+}SeekMode;
```
Comment (AryanSalmanpour) [inline utils/video_demuxer.h]: add a space `}SeekMode` -> `} SeekMode`

File: utils/video_demuxer.h [hunk 3]
```diff
@@ -40,6 +40,84 @@ extern "C" {
...
+
+/**
+ * @brief Enum for Seek Criteria
+ * 
+ */
+typedef enum SeekCriteria_enum {
```
Comment (AryanSalmanpour) [inline utils/video_demuxer.h]: change `SeekCriteria_enum` to `SeekCriteriaEnum`

File: utils/video_demuxer.h [hunk 4]
```diff
@@ -40,6 +40,84 @@ extern "C" {
...
+    SEEK_CRITERIA_FRAME_NUM = 0,
+    SEEK_CRITERIA_TIME_STAMP = 1,
+    SEEK_CRITERIA_NUM,
+}SeekCriteria;
```
Comment (AryanSalmanpour) [inline utils/video_demuxer.h]: add a space  `}SeekCriteria` -> `} SeekCriteria`

File: utils/video_demuxer.h [hunk 5]
```diff
@@ -40,6 +40,84 @@ extern "C" {
...
+    VideoSeekContext()
+        : use_seek_(false), seek_frame_(0), seek_mode_(SEEK_MODE_PREV_KEY_FRAME), seek_crit_(SEEK_CRITERIA_FRAME_NUM),
+        out_frame_pts_(0), out_frame_duration_(0), num_frames_decoded_(0U) {}
+
+    VideoSeekContext(uint64_t frame_id)
+        : use_seek_(true), seek_frame_(frame_id), seek_mode_(SEEK_MODE_PREV_KEY_FRAME),
+        seek_crit_(SEEK_CRITERIA_FRAME_NUM), out_frame_pts_(0), out_frame_duration_(0), num_frames_decoded_(0U) {}
+
+    VideoSeekContext& operator=(const VideoSeekContext& other)
+    {
```
Comment (AryanSalmanpour) [inline utils/video_demuxer.h]: move the opening bracket to the previous line.

File: utils/video_demuxer.h [hunk 6]
```diff
@@ -54,13 +132,29 @@ class VideoDemuxer {
...
+            // Convert integer timestamp representation to AV_TIME_BASE and switch to fixed_point
+            auto const ts_tbu = llround(ts_sec * AV_TIME_BASE);
+
+            // Rescale the timestamp to value represented in stream base units;
+            AVRational time_factor = {1, AV_TIME_BASE};
+            return av_rescale_q(ts_tbu, time_factor, av_fmt_input_ctx_->streams[av_stream_]->time_base);
+        }
+
+        int64_t TsFromFrameNumber(int64_t frame_num) {
+            auto const ts_sec = (double)frame_num / frame_rate_;
```
Comment (AryanSalmanpour) [inline utils/video_demuxer.h]: use `static_cast<double>(frame_num)` instead of `(double)frame_num`

File: utils/video_demuxer.h [hunk 7]
```diff
@@ -197,7 +293,10 @@ VideoDemuxer::VideoDemuxer(AVFormatContext *av_fmt_input_ctx) : av_fmt_input_ctx
     height_ = av_fmt_input_ctx_->streams[av_stream_]->codecpar->height;
     chroma_format_ = (AVPixelFormat)av_fmt_input_ctx_->streams[av_stream_]->codecpar->format;
     bit_rate_ = av_fmt_input_ctx_->streams[av_stream_]->codecpar->bit_rate;
-    frame_rate_ = av_fmt_input_ctx_->streams[av_stream_]->r_frame_rate;
+    if (av_fmt_input_ctx_->streams[av_stream_]->r_frame_rate.den != 0)
+        frame_rate_ = static_cast<double> (av_fmt_input_ctx_->streams[av_stream_]->r_frame_rate.num) / static_cast<double> (av_fmt_input_ctx_->streams[av_stream_]->r_frame_rate.den);
```
Comment (AryanSalmanpour) [inline utils/video_demuxer.h]: remove extra space between `>` and `(` in `static_cast<double> (av_fmt_input_ctx_`. Please fix all several other occurrences of the same issue in this file.

File: utils/video_demuxer.h [hunk 8]
```diff
@@ -293,6 +395,132 @@ VideoDemuxer::VideoDemuxer(AVFormatContext *av_fmt_input_ctx) : av_fmt_input_ctx
...
+    }
+
+    // Seek for single frame;
+    auto seek_frame = [&](VideoSeekContext const& seek_ctx, int flags) {
+        bool seek_backward = true;
+        int64_t timestamp = 0;
+        int ret = 0;
+
+        switch (seek_ctx.seek_crit_) {
+        case SEEK_CRITERIA_FRAME_NUM:
```
Comment (AryanSalmanpour) [inline utils/video_demuxer.h]: use 4 steps indentation for all `case`

File: utils/video_demuxer.h [hunk 9]
```diff
@@ -293,6 +395,132 @@ VideoDemuxer::VideoDemuxer(AVFormatContext *av_fmt_input_ctx) : av_fmt_input_ctx
...
+            throw std::runtime_error("ERROR: seeking for frame: ");
+        }
+    };
+
+    // Check if frame satisfies seek conditions;
+    auto is_seek_done = [&](PacketData& pkt_data, VideoSeekContext const& seek_ctx) {
+        int64_t target_ts = 0;
+
+        switch (seek_ctx.seek_crit_) {
+        case SEEK_CRITERIA_FRAME_NUM:
```
Comment (AryanSalmanpour) [inline utils/video_demuxer.h]: use 4 steps indentation for all `case`

**Review by AryanSalmanpour — CHANGES_REQUESTED** [review]
@rrawther added a few minor formatting issues. Please take a look. Thanks.

File: utils/video_demuxer.h (hunk 6 above)
Comment (rrawther) [reply utils/video_demuxer.h]: done

**Review by AryanSalmanpour — APPROVED** [review]

**Pull request merged**